django_setup_configuration/config_settings.py

```diff
@@ -0,0 +1,227 @@
+from dataclasses import dataclass
+from typing import Mapping, Sequence, Type
+
+from django.conf import settings
+from django.contrib.postgres.fields import ArrayField
+from django.db import models
+from django.db.models.fields import NOT_PROVIDED
+from django.db.models.fields.json import JSONField
+from django.db.models.fields.related import ForeignKey, OneToOneField
+from django.utils.module_loading import import_string
+
+from .constants import basic_field_descriptions
+from .exceptions import ImproperlyConfigured
+
+
+@dataclass(frozen=True, slots=True)
+class ConfigField:
+    name: str
+    verbose_name: str
+    description: str
+    default_value: str
+    field_description: str
+
+
+class ConfigSettings:
+    """
+    Settings for configuration steps, used to generate documentation.
+
+    Attributes:
+        namespace (`str`): the namespace of configuration variables for a given
+            configuration
+        file_name (`str`): the name of the file where the documentation is stored
+        models (`list`): a list of models from which documentation is retrieved
+        update_field_descriptions (`bool`): if `True`, custom model fields
+            (along with their descriptions) are loaded via the settings variable
+            `DJANGO_SETUP_CONFIG_CUSTOM_FIELDS`
+        required_settings (`list`): required settings for a configuration step
+        optional_settings (`list`): optional settings for a configuration step
+        detailed_info (`dict`): information for configuration settings which are
+            not associated with a particular model field
+
+    Example:
+        Given a configuration step `FooConfigurationStep`: ::
+
+        FooConfigurationStep(BaseConfigurationStep):
+            verbose_name = "Configuration step for Foo"
+            enable_setting = "FOO_CONFIG_ENABLE"
+            config_settings = ConfigSettings(
+                namespace="FOO",
+                file_name="foo",
+                models=["FooConfigurationModel"],
+                required_settings=[
+                    "FOO_SOME_SETTING",
+                    "FOO_SOME_OTHER_SETTING",
+                ],
+                optional_settings=[
+                    "FOO_SOME_OPT_SETTING",
+                    "FOO_SOME_OTHER_OPT_SETTING",
+                ],
+                detailed_info={
+                    "example_non_model_field": {
+                        "variable": "FOO_EXAMPLE_NON_MODEL_FIELD",
+                        "description": "Documentation for a field that could not
+                            be retrievend from a model",
+                        "possible_values": "string (URL)",
+                    },
+                },
+            )
+    """
+
+    namespace: str
+    file_name: str
+    models: list[Type[models.Model]] | None
+    required_settings: list[str] = []
+    optional_settings: list[str] = []
```

Comment:
I'm not sure if it's a real problem here, but it's generally not a good idea to use lists/dicts as default arguments, because if they are mutated anywhere, the actual default changes for new instantiations of the class as well:

```python
class Foo:
    models: list = []

    def __init__(self, *args, **kwargs):
        for key, value in kwargs.items():
            setattr(self, key, value)
        ...
        self.models.append(10)

obj = Foo()
print(obj.models)

obj2 = Foo()
print(obj2.models)
```
Results in:
```
[10]
[10, 10]
```